# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/script/ScriptLoadRequest.cpp
# Commit: ea775927f4ab
# Full Hash: ea775927f4ab98bb722ed493061868ad917af829
# Author: Yulia Startsev <ystartsev@mozilla.com>
# Date: 2022-02-18 21:52:29
# Regressor Bug: 1742437
# File Overlap Count: 1
# Description:
#   Bug 1742437 - move mWebExtGlobal from ScriptLoadRequest to DOMScriptLoadContext; r=jonco,smaug
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D136744
# ==============================================================================

diff -r c33b42cabd55 -r ea775927f4ab dom/script/ScriptLoadRequest.cpp
--- a/dom/script/ScriptLoadRequest.cpp	Fri Feb 18 12:45:49 2022 +0000
+++ b/dom/script/ScriptLoadRequest.cpp	Fri Feb 18 12:45:49 2022 +0000
@@ -32,20 +32,17 @@
 // ScriptFetchOptions
 //////////////////////////////////////////////////////////////
 
-NS_IMPL_CYCLE_COLLECTION(ScriptFetchOptions, mTriggeringPrincipal,
-                         mWebExtGlobal)
+NS_IMPL_CYCLE_COLLECTION(ScriptFetchOptions, mTriggeringPrincipal)
 
 NS_IMPL_CYCLE_COLLECTION_ROOT_NATIVE(ScriptFetchOptions, AddRef)
 NS_IMPL_CYCLE_COLLECTION_UNROOT_NATIVE(ScriptFetchOptions, Release)
 
 ScriptFetchOptions::ScriptFetchOptions(mozilla::CORSMode aCORSMode,
                                        ReferrerPolicy aReferrerPolicy,
-                                       nsIPrincipal* aTriggeringPrincipal,
-                                       nsIGlobalObject* aWebExtGlobal)
+                                       nsIPrincipal* aTriggeringPrincipal)
     : mCORSMode(aCORSMode),
       mReferrerPolicy(aReferrerPolicy),
-      mTriggeringPrincipal(aTriggeringPrincipal),
-      mWebExtGlobal(aWebExtGlobal) {
+      mTriggeringPrincipal(aTriggeringPrincipal) {
   MOZ_ASSERT(mTriggeringPrincipal);
 }
 
@@ -109,7 +106,7 @@
 NS_IMPL_CYCLE_COLLECTION_CLASS(DOMScriptLoadContext)
 
 NS_IMPL_CYCLE_COLLECTION_UNLINK_BEGIN(DOMScriptLoadContext)
-  NS_IMPL_CYCLE_COLLECTION_UNLINK(mElement)
+  NS_IMPL_CYCLE_COLLECTION_UNLINK(mElement, mWebExtGlobal)
   // XXX missing mLoadBlockedDocument ?
   if (Runnable* runnable = tmp->mRunnable.exchange(nullptr)) {
     runnable->Release();
@@ -118,14 +115,16 @@
 NS_IMPL_CYCLE_COLLECTION_UNLINK_END
 
 NS_IMPL_CYCLE_COLLECTION_TRAVERSE_BEGIN(DOMScriptLoadContext)
-  NS_IMPL_CYCLE_COLLECTION_TRAVERSE(mLoadBlockedDocument, mRequest, mElement)
+  NS_IMPL_CYCLE_COLLECTION_TRAVERSE(mLoadBlockedDocument, mRequest, mElement,
+                                    mWebExtGlobal)
 NS_IMPL_CYCLE_COLLECTION_TRAVERSE_END
 
 NS_IMPL_CYCLE_COLLECTION_TRACE_BEGIN(DOMScriptLoadContext)
 NS_IMPL_CYCLE_COLLECTION_TRACE_END
 
 DOMScriptLoadContext::DOMScriptLoadContext(Element* aElement,
-                                           ScriptLoadRequest* aRequest)
+                                           ScriptLoadRequest* aRequest,
+                                           nsIGlobalObject* aWebExtGlobal)
     : mScriptMode(ScriptMode::eBlocking),
       mScriptFromHead(false),
       mIsInline(true),
@@ -140,6 +139,7 @@
       mRunnable(nullptr),
       mIsPreload(false),
       mElement(aElement),
+      mWebExtGlobal(aWebExtGlobal),
       mRequest(aRequest),
       mUnreportedPreloadError(NS_OK) {}
 
@@ -280,17 +280,26 @@
 bool DOMScriptLoadContext::IsPreload() const {
   if (mRequest->IsModuleRequest() && !mRequest->IsTopLevel()) {
     ModuleLoadRequest* root = mRequest->AsModuleRequest()->GetRootModule();
-    return root->mLoadContext->IsPreload();
+    return root->GetLoadContext()->IsPreload();
   }
 
   MOZ_ASSERT_IF(mIsPreload, !GetScriptElement());
   return mIsPreload;
 }
 
+nsIGlobalObject* DOMScriptLoadContext::GetWebExtGlobal() const {
+  if (mRequest->IsModuleRequest() && !mRequest->IsTopLevel()) {
+    ModuleLoadRequest* root = mRequest->AsModuleRequest()->GetRootModule();
+    return root->GetLoadContext()->GetWebExtGlobal();
+  }
+
+  return mWebExtGlobal;
+}
+
 nsIScriptElement* DOMScriptLoadContext::GetScriptElement() const {
   if (mRequest->IsModuleRequest() && !mRequest->IsTopLevel()) {
     ModuleLoadRequest* root = mRequest->AsModuleRequest()->GetRootModule();
-    return root->mLoadContext->GetScriptElement();
+    return root->GetLoadContext()->GetScriptElement();
   }
   nsCOMPtr<nsIScriptElement> scriptElement = do_QueryInterface(mElement);
   return scriptElement;